pkg/block/s3/adapter.go

```diff
@@ -545,45 +479,108 @@ func (a *Adapter) copyPart(ctx context.Context, sourceObj, destinationObj block.
 		return nil, err
 	}
 
-	bucket, key, qualifiedKey, err := a.extractParamsFromObj(destinationObj)
+	bucket, key, _, err := a.extractParamsFromObj(destinationObj)
 	if err != nil {
 		return nil, err
 	}
 
 	uploadPartCopyObject := s3.UploadPartCopyInput{
 		Bucket:     aws.String(bucket),
 		Key:        aws.String(key),
-		PartNumber: aws.Int64(int64(partNumber)),
+		PartNumber: int32(partNumber),
 		UploadId:   aws.String(uploadID),
 		CopySource: aws.String(fmt.Sprintf("%s/%s", srcKey.GetStorageNamespace(), srcKey.GetKey())),
 	}
 	if byteRange != nil {
 		uploadPartCopyObject.CopySourceRange = byteRange
 	}
-	client := a.clients.Get(ctx, qualifiedKey.GetStorageNamespace())
-	req, resp := client.UploadPartCopyRequest(&uploadPartCopyObject)
-	req.SetContext(ctx)
-	err = req.Send()
+	client := a.clients.Get(ctx, bucket)
+	resp, err := client.UploadPartCopy(ctx, &uploadPartCopyObject)
 	if err != nil {
 		return nil, err
 	}
 	if resp == nil || resp.CopyPartResult == nil || resp.CopyPartResult.ETag == nil {
 		return nil, ErrMissingETag
 	}
+
 	etag := strings.Trim(*resp.CopyPartResult.ETag, `"`)
-	// x-amz-server-side-* headers
-	headers := make(http.Header)
-	for k, v := range req.HTTPResponse.Header {
-		if strings.HasPrefix(k, "X-Amz-Server-Side-") {
-			headers[k] = v
-		}
-	}
 	return &block.UploadPartResponse{
 		ETag:             etag,
-		ServerSideHeader: headers,
+		ServerSideHeader: extractSSHeaderUploadPartCopy(resp),
 	}, nil
 }
 
+// extractAmzServerSideHeaders extracts the x-amz-server-side-* headers from the response
+// TODO(barak): verify that this is the correct way to extract the headers
```

Comment:
1. TODO needs to be done and removed.
2. Move these extractors to another file.

Reply:
Moved to a differnt file - the use of these headers tested by get/put object on an encrypted bucket.